zerver/lib/push_notifications.py

```diff
@@ -610,7 +610,20 @@ def send_notifications_to_bouncer(
         "apple_devices": [device.token for device in apple_devices],
     }
     # Calls zilencer.views.remote_server_notify_push
-    response_data = send_json_to_push_bouncer("POST", "push/notify", post_data)
+
+    try:
+        response_data = send_json_to_push_bouncer("POST", "push/notify", post_data)
+    except PushNotificationsDisallowedByBouncerError as e:
+        logger.error("Bouncer refused to send push notification: %s", e.reason)
+        do_set_realm_property(
+            user_profile.realm,
+            "push_notifications_enabled",
+            False,
+            acting_user=None,
+        )
+        do_set_push_notifications_enabled_end_timestamp(user_profile.realm, None, acting_user=None)
```

Comment:
We didn't discuss doing these, but I think it's correct and useful probably?

Reply:
Yeah, though we should probably then add a commit to also do the opposite if the server did allow us to send the notification.

Reply:
That should already trigger, via the the mechanism right below I think:

https://github.com/zulip/zulip/blob/94558478dff538affbb5f4b22ca2e1d6dc184e8b/zerver/lib/push_notifications.py#L647-L660

Reply:
@timabbott I added a chunk of code to my test in the 2nd commit for verifying this is indeed the case.

Reply:
Great, thanks! Yeah I didn't expand the context.